src/hp/derp/client.rs

```diff
@@ -0,0 +1,515 @@
+        if let ReceivedMessage::ServerInfo {
+            token_bucket_bytes_per_second,
+            ..
+        } = sm
+        {
+            if token_bucket_bytes_per_second == 0 {
+                self.rate_limiter = None;
+            } else {
+                // make a new rate_limiter & add it to the client
+                todo!("the rate limiter stuff is a mess, figure it out");
+            }
+        }
+    }
+
+    async fn local_addr(&self) -> Result<SocketAddr> {
+        {
+            let recv_err = self.recv_err.lock().await;
+            if let Some(e) = &*recv_err {
+                bail!(e.clone());
+            }
+        }
+        Ok(self.conn.local_addr()?)
+    }
+
+    /// Reads a messages from a DERP server.
+    ///
+    /// The returned message may alias memory owned by the [`Client`]; if
+    /// should only be accessed until the next call to [`Client`].
+    ///
+    /// Once [`recv`] returns an error, the [`Client`] is dead forever.
+    pub async fn recv(&self) -> Result<ReceivedMessage> {
+        self.recv_check_error(Duration::from_secs(120)).await
+    }
+
+    async fn recv_check_error(&self, timeout: Duration) -> Result<ReceivedMessage> {
+        {
+            let recv_err = self.recv_err.lock().await;
+            if let Some(err) = &*recv_err {
+                bail!(err.clone());
+            }
+        }
+        match self.recv_timeout(timeout).await {
+            Ok(m) => Ok(m),
+            Err(e) => {
+                let mut recv_err = self.recv_err.lock().await;
+                // if it's errored on a simultaneous call to `recv` alread, just
+                // return the other error
+                if let Some(err) = &*recv_err {
+                    bail!(err.clone());
+                } else {
+                    *recv_err = Some(e.to_string());
+                    bail!(e);
+                }
+            }
+        }
+    }
+
+    async fn recv_timeout(&self, timeout: Duration) -> Result<ReceivedMessage> {
+        todo!();
+    }
+
+    async fn recv_0(&mut self) -> Result<ReceivedMessage> {
+        let mut frame_payload = BytesMut::with_capacity(4 * 1024);
+        loop {
+            let (frame_type, frame_len) = read_frame_header(&mut self.reader).await?;
+            if frame_len as usize > MAX_FRAME_SIZE {
+                bail!("unexpectedly large frame of {} bytes returned", frame_len);
+            }
+            let mut read_total = 0;
+            loop {
+                let read = self.reader.read(&mut frame_payload).await?;
+                if read == 0 {
+                    break;
+                }
+                read_total += read;
+            }
+            if read_total != frame_len as usize {
+                bail!(
+                    "unexpected number of bytes sent in frame, said {}, received {}",
+                    frame_len,
+                    read_total
+                );
+            }
```

Comment:
aren't you re-implementing `read_exact()`?  if you create the frame_payload after you know frame_len you might be able to use that instead?

Reply:
yes, I did... I basically couldn't figure out how to specify the exact length of the buffer I wanted to use... until just now realizing I could pass a slice of a specific size.

Thanks!